clang/lib/3C/ConstraintVariables.cpp

```diff
@@ -1678,14 +1683,19 @@ void FunctionVariableConstraint::brainTransplant(ConstraintVariable *FromCV,
 }
 
 void FunctionVariableConstraint::mergeDeclaration(ConstraintVariable *FromCV,
-                                                  ProgramInfo &I) {
+                                                  ProgramInfo &I,
+                                                  std::string &ReasonFailed) {
   // `this`: is the declaration the tool saw first
   // `FromCV`: is the declaration seen second, it cannot have defered constraints
   FVConstraint *From = dyn_cast<FVConstraint>(FromCV);
   assert(From != nullptr);
   assert(From->getDeferredParams().size() == 0);
   // Transplant returns.
-  ReturnVar->mergeDeclaration(From->ReturnVar, I);
+  // Our first sanity check is to ensure the function's returns type-check
+  std::string Base = ReasonFailed = "for return value";
+  ReturnVar->mergeDeclaration(From->ReturnVar, I, ReasonFailed);
+  if (ReasonFailed != Base) return;
+  ReasonFailed = "";
```

Comment:
Why is this here? I would think that if you start with `ReasonFailed` as `""` at the very beginning, then it will just stay that way. You shouldn't have to reset it. Same with the case below.

Reply:
Yes, you're right. The reason I reset it here though is because if you take a look at 1695, I set ReasonFailed to be "for return value." I do this so that if we ever error out, the error message will be descriptive enough to indicate that the return value is the culprit here. If I didn't reset ReasonFailed to be "", then when we pop back up, then `insertIntoExternalFunctionMap` will detect that the string has changed (from "" to "for return value") and then issue an error, when an error really shouldn't be fired. It's admittedly very hacky and is a direct result of me trying to multi-purpose this variable as a boolean and as a string, so if we think it's cleaner to treat them separately, I can make that change.

Reply:
I don't follow this. Couldn't you set the error message in 1697? I.e., change the above code to
```
ReturnVar->mergeDeclaration(From->ReturnVar, I, ReasonFailed);
if (ReasonFailed != "") {
  ReasonFailed += "for return value"; 
  return;
}
```
or something like that? I worry that if you are resetting the value in various places that (a) you are confusing the algorithm a little bit, but more importantly (b) you might cause a failure to get inadvertently missed.

Reply:
Yes, this seems much better. Fixed in most recent commit (yet to push, since I want to report back on changing the error to warning below)